// azure-kusto-data/dist-esm/src/clientDetails.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientDetails = void 0;
const core_util_1 = require("@azure/core-util");
const os_1 = require("os");
const version_1 = require("./version");
// REPLACE_REGEX = re.compile(r"[\r\n\s{}|]+")
const ReplaceRegex = /[\r\n\s{}|]+/g;
const None = "[none]";
class ClientDetails {
    constructor(applicationNameForTracing, userNameForTracing) {
        this.applicationNameForTracing = applicationNameForTracing;
        this.userNameForTracing = userNameForTracing;
        if (this.applicationNameForTracing === null) {
            this.applicationNameForTracing = ClientDetails.defaultApplication();
        }
        if (this.userNameForTracing === null) {
            this.userNameForTracing = ClientDetails.defaultUser();
        }
        this.versionForTracing = ClientDetails.defaultVersion();
    }
    static defaultApplication() {
        var _a, _b;
        if (core_util_1.isNode) {
            return ((_a = process.env) === null || _a === void 0 ? void 0 : _a.npm_package_name) || process.title || None;
        }
        else {
            return ((_b = window === null || window === void 0 ? void 0 : window.location) === null || _b === void 0 ? void 0 : _b.href) || None;
        }
    }
    static defaultUser() {
        var _a;
        if (core_util_1.isNode) {
            let username;
            try {
                username = (0, os_1.userInfo)().username;
            }
            catch (err) {
                /* Ignore possible errors like "uv_os_get_passwd returned ENOMEM" that may occur in some environments. */
                // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
                if (((_a = err.info) === null || _a === void 0 ? void 0 : _a.code) !== "ENOMEM") {
                    throw err;
                }
            }
            return username || (process.env.USERDOMAIN ? `${process.env.USERDOMAIN}\\${process.env.USERNAME}` : process.env.USERNAME) || None;
        }
        else {
            return None;
        }
    }
    static defaultVersion() {
        var _a;
        return this.formatHeader([
            ["Kusto.JavaScript.Client", version_1.SDK_VERSION],
            ["Runtime." + (core_util_1.isNode ? "Node" : "Browser"), (core_util_1.isNode ? process.version : (_a = window === null || window === void 0 ? void 0 : window.navigator) === null || _a === void 0 ? void 0 : _a.userAgent) || None],
        ]);
    }
    static escapeHeader(header) {
        return `{${header.replace(ReplaceRegex, "_")}}`;
    }
    static formatHeader(args) {
        return args
            .filter(([key, val]) => key && val)
            .map(([key, val]) => `${key}:${this.escapeHeader(val)}`)
            .join("|");
    }
    static setConnectorDetails(name, version, app_name = null, app_version = null, send_user = false, override_user = null, additional_fields = null) {
        const params = [["Kusto." + name, version]];
        app_name = app_name || this.defaultApplication();
        app_version = app_version || None;
        params.push(["App." + this.escapeHeader(app_name), app_version]);
        params.push(...(additional_fields || []));
        let user = None;
        if (send_user) {
            user = override_user || this.defaultUser();
        }
        return new ClientDetails(this.formatHeader(params), user);
    }
    getHeaders() {
        return {
            "x-ms-client-version": this.versionForTracing,
            "x-ms-app": this.applicationNameForTracing,
            "x-ms-user": this.userNameForTracing,
        };
    }
}
exports.ClientDetails = ClientDetails;
